File: packages/elixir/lumis/test/conformance_test.exs
defmodule Lumis.ConformanceTest do
  use ExUnit.Case, async: true
  @moduletag :conformance

  @conformance_dir Path.expand("../../../../fixtures/conformance", __DIR__)

  @fixture_names @conformance_dir
                 |> Path.join("*/fixture.json")
                 |> Path.wildcard()
                 |> Enum.map(&Path.dirname/1)
                 |> Enum.map(&Path.basename/1)
                 |> Enum.sort()

  # The same nine the JavaScript conformance suite declares. Highlighting would
  # load them anyway; naming them keeps this suite comparing formatter output
  # rather than measuring a first load, and matches what the browser has to do,
  # since it cannot load inside the walk.
  setup_all do
    :ok =
      Lumis.Languages.load(~w(
        json html javascript css lua markdown markdown_inline mdx python
      ))

    :ok
  end

  defp load_fixture(name) do
    dir = Path.join(@conformance_dir, name)
    metadata = dir |> Path.join("fixture.json") |> File.read!() |> Jason.decode!()

    Map.merge(metadata, %{
      "source" => File.read!(Path.join(dir, "source.txt")),
      "htmlInline" => File.read!(Path.join(dir, "html-inline.html")),
      "htmlLinked" => File.read!(Path.join(dir, "html-linked.html")),
      "htmlMultiThemes" => File.read!(Path.join(dir, "html-multi-themes.html")),
      "bbcode" => File.read!(Path.join(dir, "bbcode.txt")),
      "terminal" => File.read!(Path.join(dir, "terminal.txt"))
    })
  end

  defp rainbow_brackets(fixture), do: fixture["rainbowBrackets"] || false

  for name <- @fixture_names do
    describe name do
      @tag fixture: name
      test "html_inline" do
        fixture = load_fixture(unquote(name))

        assert Lumis.highlight!(fixture["source"],
                 formatter: {
                   :html_inline,
                   language: fixture["language"],
                   theme: fixture["theme"],
                   rainbow_brackets: rainbow_brackets(fixture)
                 }
               ) == fixture["htmlInline"]
      end

      @tag fixture: name
      test "html_linked" do
        fixture = load_fixture(unquote(name))

        assert Lumis.highlight!(fixture["source"],
                 formatter:
                   {:html_linked,
                    language: fixture["language"], rainbow_brackets: rainbow_brackets(fixture)}
               ) == fixture["htmlLinked"]
      end

      @tag fixture: name
      test "html_multi_themes" do
        fixture = load_fixture(unquote(name))

        assert Lumis.highlight!(fixture["source"],
                 formatter: {
                   :html_multi_themes,
                   language: fixture["language"],
                   themes: [main: fixture["theme"]],
                   default_theme: "main",
                   rainbow_brackets: rainbow_brackets(fixture)
                 }
               ) == fixture["htmlMultiThemes"]
      end

      @tag fixture: name
      test "terminal" do
        fixture = load_fixture(unquote(name))

        assert Lumis.highlight!(fixture["source"],
                 formatter: {
                   :terminal,
                   language: fixture["language"],
                   theme: fixture["theme"],
                   rainbow_brackets: rainbow_brackets(fixture)
                 }
               ) == fixture["terminal"]
      end

      @tag fixture: name
      test "bbcode_scoped" do
        fixture = load_fixture(unquote(name))

        assert Lumis.highlight!(fixture["source"],
                 formatter:
                   {:bbcode_scoped,
                    language: fixture["language"], rainbow_brackets: rainbow_brackets(fixture)}
               ) == fixture["bbcode"]
      end
    end
  end
end

File: packages/javascript/lumis/test/conformance.test.ts
import { beforeAll, describe, expect, it } from "vitest";

import dracula from "../../themes/dist/json/dracula.json";
import css from "../langs/css.ts";
import html from "../langs/html.ts";
import javascript from "../langs/javascript.ts";
import json from "../langs/json.ts";
import lua from "../langs/lua.ts";
import markdown from "../langs/markdown.ts";
import markdownInline from "../langs/markdown_inline.ts";
import mdx from "../langs/mdx.ts";
import python from "../langs/python.ts";
import rust from "../langs/rust.ts";
import { createHighlighter } from "../src/index.js";
import {
  bbcodeScoped,
  htmlInline,
  htmlLinked,
  htmlMultiThemes,
  terminal,
} from "../src/formatters.js";
import type { Highlighter } from "../src/core/highlighter.js";
import type { Language, Theme } from "../src/types.js";

import { loadConformanceFixtures } from "./conformance.js";
import { configureLocalWasmResolver } from "./wasm.js";

const conformanceFixtures = loadConformanceFixtures();
const theme: Theme = dracula;

const langBundles: Record<string, Language> = {
  json,
  html,
  javascript,
  css,
  lua,
  markdown,
  markdownInline,
  mdx,
  python,
  rust,
};

function getLanguage(id: string): Language {
  const language = langBundles[id];
  if (!language) throw new Error(`No bundle for language "${id}" in test setup`);
  return language;
}

let highlighter: Highlighter;

beforeAll(async () => {
  configureLocalWasmResolver([
    "diff",
    "json",
    "html",
    "javascript",
    "css",
    "lua",
    "markdown",
    "markdown_inline",
    "python",
    "rust",
  ]);
  // No `wasm` overrides. Parser bytes and the package that describes them have
  // to come from the same place, or the integrity check rejects the pair: CI
  // stages freshly built parsers while `ensureLocalWasm` returns the committed
  // fixture, which is a different build of the same grammar.
  highlighter = await createHighlighter({
    languages: [json, html, javascript, css, lua, markdown, markdownInline, mdx, python, rust],
  });
}, 120_000);

describe.each(conformanceFixtures.map((f) => [f.name, f]))("%s", (_name, fixture) => {
  it("html-inline", () => {
    const output = highlighter.highlight(
      fixture.source,
      htmlInline({
        language: getLanguage(fixture.language),
        theme,
        rainbowBrackets: fixture.rainbowBrackets,
      }),
    );
    expect(output).toBe(fixture.htmlInline);
  });

  it("html-linked", () => {
    const output = highlighter.highlight(
      fixture.source,
      htmlLinked({
        language: getLanguage(fixture.language),
        rainbowBrackets: fixture.rainbowBrackets,
      }),
    );
    expect(output).toBe(fixture.htmlLinked);
  });

  it("html-multi-themes", () => {
    const output = highlighter.highlight(
      fixture.source,
      htmlMultiThemes({
        language: getLanguage(fixture.language),
        themes: { main: theme },
        defaultTheme: "main",
        rainbowBrackets: fixture.rainbowBrackets,
      }),
    );
    expect(output).toBe(fixture.htmlMultiThemes);
  });

  it("bbcodeScoped", () => {
    const output = highlighter.highlight(
      fixture.source,
      bbcodeScoped({
        language: getLanguage(fixture.language),
        rainbowBrackets: fixture.rainbowBrackets,
      }),
    );
    expect(output).toBe(fixture.bbcode);
  });

  it("terminal", () => {
    const output = highlighter.highlight(
      fixture.source,
      terminal({
        language: getLanguage(fixture.language),
        theme,
        rainbowBrackets: fixture.rainbowBrackets,
      }),
    );
    expect(output).toBe(fixture.terminal);
  });
});
